api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java

```diff
@@ -230,6 +233,42 @@ public JSONObject getNotebookContents(String bucketName, String notebookName) {
     return cloudStorageClient.readBlobAsJson(blob);
   }
 
+  @Override
+  public KernelTypeEnum getNotebookKernel(JSONObject notebookFile) {
+    try {
+      String language =
+          Optional.of(notebookFile.getJSONObject("metadata"))
+              .flatMap(metaDataObj -> Optional.of(metaDataObj.getJSONObject("kernelspec")))
+              .map(kernelSpec -> kernelSpec.getString("language"))
+              .orElse("Python");
+
+      if ("R".equals(language)) {
+        return KernelTypeEnum.R;
+      } else {
+        return KernelTypeEnum.PYTHON;
+      }
+    } catch (JSONException e) {
+      // If we can't find metadata to parse, default to python.
+      return KernelTypeEnum.PYTHON;
+    }
+  }
+
+  @Override
+  public KernelTypeEnum getNotebookKernel(
+      String workspaceNamespace, String workspaceName, String notebookName) {
+    workspaceAuthService.enforceWorkspaceAccessLevel(
+        workspaceNamespace, workspaceName, WorkspaceAccessLevel.READER);
+
+    String bucketName =
+        fireCloudService
+            .getWorkspace(workspaceNamespace, workspaceName)
+            .getWorkspace()
+            .getBucketName();
+
+    Blob blob = getBlobWithSizeConstraint(bucketName, notebookName);
```

Comment:
Yep. I had considered this case but decided to keep the size restriction because that's what the old generate code did. I don't have an error msg specific to this case but the component will gracefully handle all failed API calls with at least somewhat customized message (depending on which call failed).